ops/framework.py

```diff
@@ -954,6 +954,14 @@ def _reemit(self, single_event_path: Optional[str] = None):
                             # Regular call to the registered method.
                             custom_handler(event)
 
+            else:
+                logger.warning(
+                    f"Weakref to Object at path {observer_path} has been killed somewhere "
+                    "between charm initialization and event emission. "
+                    "This can mean that you are using an Object which is not hard-referenced "
+                    "anywhere from the CharmBase tree. We won't be able to emit any events on it."
```

Comment:
Suggestion to tighten up the message and make it a bit more accurate:

```suggestion
                    f"Reference to ops.Object at path {observer_path} has been garbage collected "
                    "between when the charm was initialised and when the event was emitted. "
                    "Make sure sure you store a reference to the observer."
```